src/sage/rings/semirings/tropical_variety.py

```diff
@@ -0,0 +1,1124 @@
+        sage: p1 = R(7) + T(4)*x + y + R(4)*x*y + R(3)*y^2 + R(-3)*x^2
+        sage: tv = p1.tropical_variety(); tv
+        Tropical curve of (-3)*x^2 + 4*x*y + 3*y^2 + 4*x + 0*y + 7
+        sage: tv.components()
+        [[(3, t1), [t1 <= 0], 1],
+        [(-t1 + 3, t1), [0 <= t1, t1 <= 2], 1],
+        [(t1, 2), [t1 <= 1], 2],
+        [(t1, 0), [3 <= t1, t1 <= 7], 1],
+        [(7, t1), [t1 <= 0], 1],
+        [(t1 - 1, t1), [2 <= t1], 1],
+        [(t1 + 7, t1), [0 <= t1], 1]]
+        sage: tv.plot()
+        Graphics object consisting of 8 graphics primitives
+
+    .. PLOT::
+        :width: 300 px
+
+        T = TropicalSemiring(QQ, use_min=False)
+        R = PolynomialRing(T, ('x,y'))
+        x, y = R.gen(), R.gen(1)
+        p1 = R(7) + T(4)*x + y + R(4)*x*y + R(3)*y**2 + R(-3)*x**2
+        sphinx_plot(p1.tropical_variety().plot())
+
+    If the tropical polynomial have `n>2` variables, then the result will be
+    a tropical hypersurface embedded in a real space `\RR^n`::
+
+        sage: T = TropicalSemiring(QQ)
+        sage: R.<a,x,y,z> = PolynomialRing(T)
+        sage: p1 = x*y + R(-1/2)*x*z + R(4)*z^2 + a*x
+        sage: tv = p1.tropical_variety(); tv
+        Tropical hypersurface of 0*a*x + 0*x*y + (-1/2)*x*z + 4*z^2
+        sage: tv.components()
+        [[(t1, t2, t3 - 1/2, t3), [t2 - 9/2 <= t3, t3 <= t1 + 1/2, t2 - 5 <= t1], 1],
+        [(t1, 2*t2 - t3 + 4, t3, t2), [t3 + 1/2 <= t2, t3 <= t1], 1],
+        [(t1, t2, t1, t3), [max(t1 + 1/2, 1/2*t1 + 1/2*t2 - 2) <= t3], 1],
+        [(t1, t2 + 9/2, t3, t2), [t2 <= min(t3 + 1/2, t1 + 1/2)], 1],
+        [(t1 - 1/2, t2, t3, t1), [t2 - 9/2 <= t1, t1 <= t3 + 1/2, t2 - 5 <= t3], 1],
+        [(2*t1 - t2 + 4, t2, t3, t1), [t1 <= min(1/2*t2 + 1/2*t3 - 2, t2 - 9/2)], 1]]
+    """
+    def __init__(self, poly):
+        r"""
+        Initialize ``self``.
+
+        EXAMPLES::
+
+            sage: T = TropicalSemiring(QQ)
+            sage: R.<x,y> = PolynomialRing(T)
+            sage: tv = (x+y).tropical_variety()
+            sage: TestSuite(tv).run()
+
+        TESTS::
+
+            sage: from sage.rings.semirings.tropical_variety import TropicalVariety
+            sage: R.<x,y> = QQ[]
+            sage: p1 = x + y
+            sage: TropicalVariety(p1)
+            Traceback (most recent call last):
+            ...
+            ValueError: x + y is not a multivariate tropical polynomial
+        """
+        import operator
+        from itertools import combinations
+        from sage.symbolic.ring import SR
+        from sage.symbolic.relation import solve
+        from sage.arith.misc import gcd
+        from sage.rings.semirings.tropical_mpolynomial import TropicalMPolynomial
+
+        if not isinstance(poly, TropicalMPolynomial):
+            raise ValueError(f"{poly} is not a multivariate tropical polynomial")
+
+        self._poly = poly
+        self._hypersurface = []
+        if len(poly.dict()) == 1:  # constant polynomial
```

Comment:
I am pretty sure this does not test for constants, but instead monomials. Give it a try.